main/src/main/scala/sbt/Cross.scala

```diff
@@ -127,76 +128,76 @@ object Cross {
+              "not compatible with. Try issuing cross building command with tasks instead, since sbt will be able " +
+              "to ensure that cross building is only done using configured project and Scala version combinations " +
+              "that are configured."
+          )
+          state.log.debug("Scala versions configuration is:")
+          projCrossVersions.foreach {
+            case (project, versions) => state.log.debug(s"$project: $versions")
           }
+        }
 
-          // Execute using a blanket switch
-          projCrossVersions.toMap.apply(currentRef).flatMap { version =>
-            // Force scala version
-            Seq(s"$SwitchCommand $verbose $version!", aggCommand)
+        // Execute using a blanket switch
+        projCrossVersions.toMap.apply(extracted.currentRef).flatMap { version =>
+          // Force scala version
+          Seq(s"$SwitchCommand $verbose $version!", aggCommand)
+        }
+      case Right((keys, taskArgs)) =>
+        def project(key: ScopedKey[_]): Option[ProjectRef] = key.scope.project.toOption match {
+          case Some(p: ProjectRef) => Some(p)
+          case _                   => None
+        }
+        val fullArgs = if (taskArgs.trim.isEmpty) "" else s" ${taskArgs.trim}"
+        val keysByVersion = keys
+          .flatMap { k =>
+            project(k).toSeq.flatMap(crossVersions(extracted, _).map(v => v -> k))
           }
-
-        case Right(_) =>
-          // We have a key, we're likely to be able to cross build this using the per project behaviour.
-
-          // Group all the projects by scala version
-          projVersions.groupBy(_._2).mapValues(_.map(_._1)).toSeq.flatMap {
-            case (version, Seq(project)) =>
-              // If only one project for a version, issue it directly
-              Seq(s"$SwitchCommand $verbose $version $project/$aggCommand")
-            case (version, projects) if aggCommand.contains(" ") =>
-              // If the command contains a space, then the `all` command won't work because it doesn't support issuing
-              // commands with spaces, so revert to running the command on each project one at a time
-              s"$SwitchCommand $verbose $version" :: projects
-                .map(project => s"$project/$aggCommand")
-            case (version, projects) =>
-              // First switch scala version, then use the all command to run the command on each project concurrently
-              Seq(
-                s"$SwitchCommand $verbose $version",
-                projects.map(_ + "/" + aggCommand).mkString("all ", " ", "")
-              )
+          .groupBy(_._1)
+          .mapValues(_.map(_._2).toSet)
+        val commandsByVersion = keysByVersion.toSeq
+          .flatMap {
+            case (v, keys) =>
+              val projects = keys.flatMap(project)
+              keys.toSeq.flatMap { k =>
+                project(k).filter(projects.contains).flatMap { p =>
```

Comment:
The `projects.contains` seems redundant seeing as it's defined in terms of the projects in `keys` already.